tfx_addons/pandas_transform/component.py

```diff
@@ -0,0 +1,320 @@
+    Returns:
+      The corresponding tf.train.Example
+    """
+    feature = {}
+    for key, val in row.items():
+      val = [] if pd.isna(val) else [val]
+      if ptypes[key] == 'Int64':
+        feature[key] = tf.train.Feature(int64_list=tf.train.Int64List(
+            value=val))
+      elif ptypes[key] in ['float32', 'Float64']:
+        feature[key] = tf.train.Feature(float_list=tf.train.FloatList(
+            value=val))
+      elif ptypes[key] in ['string', 'object']:
+        if val != []:
+          val = [val[0].encode('utf8')]
+        feature[key] = tf.train.Feature(bytes_list=tf.train.BytesList(
+            value=val))
+      else:
+        raise ValueError('DictToExample: Type {} was unhandled'.format(
+            ptypes[key]))
+    return tf.train.Example(features=tf.train.Features(feature=feature))
+
+  def process(self, df):
+    """The Beam DoFn process function.  Yields serialized tf.train.Examples
+      Args:
+        df: A Pandas dataframe
+      Returns:
+        Yields serialized tf.train.Examples
+    """
+    ptypes = df.dtypes.apply(lambda x: x.name).to_dict()
+    for row_dict in df.to_dict(orient='records'):
+      example = self.DictToExample(row_dict, ptypes)
+      yield example.SerializeToString()
+
+if parse(tfx.__version__) >= Version('1.8.0'):
+  @component(component_annotation=system_executions.Transform, use_beam=True)
+  def PandasTransform(
+      transformed_examples: tfx.dsl.components.OutputArtifact[Examples],
+      examples: tfx.dsl.components.InputArtifact[Examples] = None,
+      schema: tfx.dsl.components.InputArtifact[Schema] = None,
+      statistics: tfx.dsl.components.InputArtifact[ExampleStatistics] = None,
+      module_file: tfx.dsl.components.Parameter[str] = None,
+      beam_pipeline: BeamComponentParameter[beam.Pipeline] = None,
+  ) -> None:
+    DoPandasTransform(examples=examples,
+                      schema=schema,
+                      statistics=statistics,
+                      transformed_examples=transformed_examples,
+                      module_file=module_file,
+                      beam_pipeline=beam_pipeline)
+else:
+  logging.info('TFX < 1.8.0')
+  logging.info('Beam custom component not supported in this version of TFX.')
+  logging.info('You might consider an upgrade.')
+  @component
+  def PandasTransform(
+      transformed_examples: tfx.dsl.components.OutputArtifact[Examples],
+      examples: tfx.dsl.components.InputArtifact[Examples] = None,
+      schema: tfx.dsl.components.InputArtifact[Schema] = None,
+      statistics: tfx.dsl.components.InputArtifact[ExampleStatistics] = None,
+      module_file: tfx.dsl.components.Parameter[str] = None) -> None:
+    DoPandasTransform(examples=examples,
+                      schema=schema,
+                      statistics=statistics,
+                      transformed_examples=transformed_examples,
+                      module_file=module_file,
+                      beam_pipeline=beam.Pipeline())
```

Comment:
nit: should we add beam_pipeline_args somehow?

Reply:
They'll be added for TFX >= 1.8.0.  I haven't really thought about how to add them for earlier versions.

Reply:
Also I really need to add more tests and at least one example.

Reply:
I added beam_pipeline_args for both the new and old versions:

>beam_pipeline_args: A string with the argv options for creating a Beam pipeline.
    Note that this is a string, not a list.  It will be split on spaces to create
    a list. If running TFX >= 1.8.0, if beam_pipeline_args are specified they will
    override the pipeline beam args.